src/deepsparse/server/build_logger.py

```diff
@@ -0,0 +1,125 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Specifies the mapping from the ServerConfig to the DeepSparse Logger
+"""
+
+from typing import Any, Dict, List, Union
+
+from deepsparse import loggers as logger_objects
+from deepsparse.loggers.config import MetricFunctionConfig
+from deepsparse.server.config import ServerConfig
+
+
+__all__ = ["build_logger"]
+
+_LOGGER_MAPPING = {"python": logger_objects.PythonLogger}
+
+
+def build_logger(server_config: ServerConfig) -> logger_objects.BaseLogger:
+    """
+    Builds a DeepSparse logger from the ServerConfig.
+
+    The process follows the tree-like hierarchy:
+
+    First: the "leaf" loggers are being built.
+    Second: if specified in ServerConfig (inside the config
+        of every server endpoint; under 'data_logging' value),
+        every "leaf" logger is wrapped inside a FunctionLogger.
+    Third: if required, the list of resulting loggers is wrapped inside a
+        MultiLogger.
+    Fourth: if specified, the resulting logger is wrapped inside the SubprocessLogger
+
+    :param server_config: the Server configuration model
+    :return: a DeepSparse logger instance
+    """
+
+    loggers_config = server_config.loggers
+    if loggers_config is None:
+        return loggers_config
+
+    leaf_loggers = build_leaf_loggers(loggers_config)
+
+    loggers = []
+    for name, leaf_logger in leaf_loggers.items():
+        target_to_metric_function_configs = _get_target_to_metric_function_configs(
+            endpoints=server_config.endpoints, logger_name=name
+        )
+        if target_to_metric_function_configs:
+            logger = logger_objects.FunctionLogger(
```

Comment:
why are we creating a separate function logger for each leaf logger? This would require us running the same function multiple times for separate leaf loggers, no?  Any reason we can't use one function logger here?

Reply:
This would require us to delegate even more functionalities to a single `FunctionLogger`. Not only "matching" and "applying the function", but also choosing which leaf logger to log (in the scenario where a metric is aiming to override the global loggers, which according to the latest news from Rob is likely to be the case).

Having to decide between scaling the number of `FunctionLoggers` linearly with the number of leaf loggers vs having even more responsibilities on the `FunctionLogger` side, I err for the former, not the latter.

Reply:
yes let's think on this, going back to multiple individual function loggers would make this a bit easier but we'd lose a bit of structure. we absolutely do not want to run the functions multiple times